### ISSUE TITLE
OpenGL backend is problematic for sprite-based games

### ISSUE THREAD

**Issue by atomnuker** [hrydgard/ppsspp#7806]
I reported this bug a while ago on IRC but I wrote it off as being something else.

Essentially, when running the OpenGL backend on any operating system, sprite based games (Rondo of Blood and Symphony of the Night from Dracula X Chronicles) will lag massively rendering them unplayable. This occurs with the current git master freshly compiled under Linux and with the official compiled binaries for Windows. On Windows, switching to the Direct9 backend will make the games run under normal speed without any slowdowns.

Switching on the "Retain changed textures" option helps a little and makes the games run at normal speed unless you move, at which point the lag is back. The textures are not retained after a fixed amount of frames has passed so it's still unplayable.

Changing the "Mode" option to anything else other than Buffered will result in a black screen for these specific games, hence they were not usefl. ALL combinations of options have been tested, including turning on hacks, none of which helped. This is clearly a problem with the OpenGL backend.

Tweaking TextureCache.cpp's macros  to allow for longer texture retain frames did not work and did not improve the performance at all.

Here are the memory card images to allow for anyone to test (put in "$configdir/PSP/SAVEDATA/ULUS10277", select Original game from the main menu and play for a bit).
http://include.strangled.net/files/ULUS10277.zip

EDIT: It appears that the DX9 backend keeps every texture forever. There's still the same type of lag when changing textures, e.g. when jumping for the first time, the textures get loaded which causes some lag. But any other jumps at any time after will not lag. OpenGL will however dump the textures after a fixed amount of frames.
Not sure if this is a bug in the DX9 backend. Either way adding an additional hack for both backends to keep all cached textures indefinitely (and probably start dropping the least used ones after the cache hits some hardcoded limit like 64 megabytes) would be nice.

**Comment by unknownbrackets** [comment]
I feel like this is an overly generic statement.

The performance behavior you're experiencing is probably not true of all games which use sprites.  For example, Tales of Eternia uses sprites and I am fairly sure it is not true of that game.  I don't have the games you mention.

Also, it may have more to do with your device or your video card drivers.  Maybe you're using Intel, with poor-at-best OpenGL drivers?  Are you sure this happens across other video cards as well, and is true for all consumers of the OpenGL backend?

"Retain changed textures" using a secondary cache for textures which change.  It applies to both OpenGL and Direct3D.  Generally, it makes performance in Popolocrois and I think one other game (both sprite based) significantly better.  In practically all other games I've tested (at least on NVIDIA cards and Adreno devices), it reduces performance or has no impact.

Direct3D 9 seems to perform better at copying regions of framebuffers on most cards.  Again, I don't know what card you're using, but usually significant impact to performance on desktops is more related to this than texture lifetime.

As an example, Star Ocean (a game I do have) might behave similar to the way you describe.  That is, when you are stationary, performance is fine.  When you move in a way that causes the screen to pan (i.e. the background scrolls), performance may drop significantly, especially on weaker cards.

The above situation in Star Ocean is **not** caused by textures at all, not even slightly.  It's because the game uses the stencil buffer to do things like z-ordering and maybe even collision detection, and uploads new data to the stencil buffer when the screen pans.  Modern GPUs through APIs like OpenGL and Direct3D don't have a quick way to upload stencil data like the PSP's GPU does.

Both backends should be decimating textures.  Also, both keep an estimate of the current "cached bytes of textures" to avoid decimating textures when it's not necessary to save VRAM (OpenGL and Direct3D textures use a lot more VRAM than PSP textures, because PSP textures typically use palettes, and OpenGL/Direct3D don't.)

If changing the constants and making it decimate less didn't help, your assumption may be incorrect.  To verify, you can change this line:

https://github.com/hrydgard/ppsspp/blob/43744b023964443946e3b19b4b20e2d8b94ae73d/GPU/GLES/TextureCache.cpp#L171

From `if (--decimationCounter_ <= 0) {` to `if (false) {`.  This will cause it to never invalidate textures.  If this solves the problem you experienced, you may want to check whether "lazy texture hashing" has an impact.  This option sometimes makes things faster, but sometimes makes things slower.

-[Unknown]

**Comment by atomnuker** [comment]
Changing the condition at line 170 does partly solve the problem. The start of SotN is playable now. Memory usage is steady and not rocketing. Rondo of Blood is also fully playable as well, at least the few minutes I played.
Once an animated health bar appears in SotN the game is back to stuttering and lagging. It's a bit faster than before, although it's still unplayable. The CPU usage is stuck at 95% so I think that might be the bottleneck here.

Lazy texture caching did neither improve or degrade performance. There was no change I could measure.

I didn't think it was the drivers' fault since the Windows and Linux versions for the Intel chips are made by 2 completely different groups with little to no contact between each other (or so I've heard). The OpenGL implementation on Windows is a different kind of a horrid mess though.

Maybe the Windows Dx9 backend is just doing things efficiently enough that my CPU isn't the bottleneck. Either way, I don't think there's probably a way to make the game playable using OpenGL without extra CPU power, so feel free to close this problem.

**Comment by hrydgard** [comment]
For this type of game that abuse the 3D pipeline to simulate the 2D pipeline of earlier consoles, the best solution might be to speed up the software renderer enough to make it usable, if OpenGL drivers don't improve their block copy performance.

**Comment by daniel229** [comment]
D3D9 is a little faster than OpenGL in my test,10% maybe.

OpenGL
![01](https://cloud.githubusercontent.com/assets/3481559/8131302/f20d5198-114b-11e5-9f6f-035e4ad47368.png)

D3D9
![02](https://cloud.githubusercontent.com/assets/3481559/8131303/f219e0e8-114b-11e5-9056-406c765c5962.png)

**Comment by LPatamon** [comment]
I have to I didn't get any slowdowns on Rondo of Blood and Symphony of the Night. I played it with the current buildbot build v1.0.1-600-gc27e634.

If something the only issue is when loading a save state, the game ends like this:

![castlevaniappsspp](https://cloud.githubusercontent.com/assets/6118529/8232090/a7ded0a6-1592-11e5-8636-e1bd51ddd945.png)

The right side gets glitched, but is nothing that gets in the way of the game. Is just the right replaced by part of the game.

Other than that, no issues, no slowdowns, character jumps all fine and such. I passed all stage 1. I don't know when the slowdowns are supposed to start, but I also played part of the stage 2 and no slowdown issues.

I noticed that @atomnuker didn't say what are his system specs. You know, CPU, RAM, Graphic Card Model and how much VRAM it has.

This issue reminded me of something I experienced on Digimon Adventure for PSP on my old 3.2 Ghz Pentium 4 HT. There were some parts of the game in which I was getting slowdowns (mostly if there were several characters on screen), but it was because my PC was too old (and yet most of that game played quite decent for me on that old PC), so maybe atomnuker's PC isn't powerful enough?
